[FEAT] Support disabling using doubled quotes to escape in CSV

ravern opened the pull request:
Closes #1507.

This PR adds the `double_quote_escape: bool` option to the `daft.read_csv` and `daft.read_csv_schema` functions. This allows the disabling of using double quotes i.e. `""` to escape quotes within a CSV file being read by Daft.

## Remaining Tasks

- [x] Write tests
- [x] Review further interfaces (I'm not sure if those two functions cover all CSV use-cases)

ravern commented:
I was hoping to get some feedback on the choice of naming the option `double_quote_escape` which could be bad(?) because it doesn't allude to the fact that it's a boolean.

---

Also, I'm having some trouble getting tests to run locally.

* `make test` passes all tests, but this is likely because `double_quote_escape` is added as a default argument and so doesn't introduce any breaking change. I also noticed this is only running the tests written in Python

* `cargo test` doesn't work due to some linker error, I've pasted the error description below, and was hoping to get some help on this.

---

```
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/Users/ravern/.rustup/toolchains/nightly-2023-08-01-aarch64-apple-darwin/lib/rustlib/aarch64-apple-darwin/bin:/Users/ravern/Repos/Eventual-Inc/Daft/venv/bin:/Users/ravern/.go/bin

...

daft_csv-1f74624d3418daf0" "-Wl,-dead_strip" "-nodefaultlibs"
  = note: ld: warning: no platform load command found in '/private/var/folders/hv/b4_thqdx1m7734k32p57nws40000gn/T/rustcvk37GV/symbols.o', assuming: macOS
          ld: Undefined symbols:
            _PyBaseObject_Type, referenced from:
                _$LT$pyo3..types..any..PyAny$u20$as$u20$pyo3..type_object..PyTypeInfo$GT$::type_object_raw::h575da96eeb8cb68e in libdaft_table-de4638bdf01a83c4.rlib[151](daft_table-de4638bdf01a83c4.y0zrou0k7joal7m.rcgu.o)
                _$LT$pyo3..types..any..PyAny$u20$as$u20$pyo3..type_object..PyTypeInfo$GT$::type_object_raw::hef75c7aad564d8eb in libdaft_dsl-0872d4c822d463eb.rlib[68](daft_dsl-0872d4c822d463eb.2pnkxy9c85hccabj.rcgu.o)
                _$LT$pyo3..pycell..PyCellBase$LT$U$GT$$u20$as$u20$pyo3..pycell..PyCellLayout$LT$T$GT$$GT$::tp_dealloc::ha80495966fbf9b04 in libcommon_io_config-16cb41f2bb8bba4e.rlib[5](common_io_config-16cb41f2bb8bba4e.18nupr6efigg4sx8.rcgu.o)

...
```

samster25 commented:
Hi @ravern,

Thanks for making a PR. I'll let @jaychia or @clarkzinzow comment on the `double_quote_escape` aspect of the question but I can help with the cargo test. With no arguments cargo will bulid with all the default features on, including python which will result in the link time error you see above. You can pass the `--no-default-features` argument to disable building with python like this: `cargo test  --package daft-csv --no-default-features -- --nocapture`

ravern commented:
I've added a unit test for this feature. This should be ready for review.

Do let me know if there should be an integration test, I couldn't find one for `has_header` so I made the assumption we don't need one here for `double_quote_escape`.

jaychia commented on src/daft-csv/test/iris_tiny_double_quote_escape.csv:
```diff
@@ -0,0 +1,22 @@
+"""sepal.""length",sepal.width,petal.length,petal.width,variety
+5.1,3.5,1.4,.2,"Setosa"
```
Let's add some actually useful escaped test-cases? E.g. `"Setosa,Sentosa"` :)

clarkzinzow replied:
+1, at a minimum it would be a good idea to add double quotes to some of the variety values (e.g. ""Setosa"") in the below rows, to ensure that the double quote escape flag works as expected for rows beyond just the header.

jaychia reviewed (commented):
Thanks for this PR @ravern - great work! Let's get @clarkzinzow to take a look since he's most familiar with the CSV APIs.

I'm also a little confused by Arrow2's multiple quoting APIs and how they might interact with each other.
1. [quote](https://docs.rs/arrow2/latest/arrow2/io/csv/read_async/struct.AsyncReaderBuilder.html#method.quote)
2. [double_quote](https://docs.rs/arrow2/latest/arrow2/io/csv/read_async/struct.AsyncReaderBuilder.html#method.double_quote)
3. [quoting](https://docs.rs/arrow2/latest/arrow2/io/csv/read_async/struct.AsyncReaderBuilder.html#method.quoting)

I wonder if a better API that may give us both control over the quote character **and** quoting behavior might be `read_csv(..., escape_quote_character: str | None = None)`, where `None` indicates that we ignore all quoting. Defaulting to `None` also gives us the best performance by default, but users have the option to fallback to the slow path for any quoting character of their choice.

cc @clarkzinzow for thoughts

clarkzinzow commented:
@jaychia I don't think we'll be able to collapse the quoting configurations into a single arg; IMO our _default_ configuration should try to conform to [RFC 4180](https://datatracker.ietf.org/doc/html/rfc4180), and I think that aligning to [pyarrow's default parsing options](https://arrow.apache.org/docs/python/generated/pyarrow.csv.ParseOptions.html#pyarrow.csv.ParseOptions
) would be a good idea as well.

Specifically, the standard says that:
1. `Each field may or may not be enclosed in double quotes. If fields are not enclosed with double quotes, then double quotes may not appear inside the fields.`
2. `Fields containing line breaks (CRLF), double quotes, and commas should be enclosed in double-quotes.`
3. `If double-quotes are used to enclose fields, then a double-quote appearing inside a field must be escaped by preceding it with another double quote.`

(1) and (2) necessitate the need for quoting in general (e.g. a string value containing a comma), and (3) details how to represent a quote in a quoted field (double quote escape). I would vote for this being the default behavior, i.e.
a. Support enclosing a field in quotes, where the quote, comma, and newline characters may not appear inside an unquoted field.
b. Use the double-quote as the default quote character.
c. To escape a quote within a quoted field, double quote it.

We'd then want to make (a), (b), and (c) configurable, but we can't represent all with a single `escape_quote_character: str | None`, since we may want to support using the single-quote as the quote character, but don't considered two successive single-quotes as an escaped single-quote in a quoted field.

My proposal would be two args, `quote_char` and `double_quote`:
- `quote_char: str | None = '"'` - A single character representing the quoting character, or None to disable quoting. Defaults to `'"'`, the double-quote.
- `double_quote: bool` - Whether two successive quotes in a quoted CSV field represent a single quote in the actual data value. Defaults to `True`.

This would adhere to RFC 4180 by default (and align with pyarrow's default semantics), while providing a path to disable quoting (to improve performance), use a different quote character (e.g. the single-quote), and disable double-quoting escapes.

jaychia commented:
Makes sense and thanks for the super in-depth explanation!

@ravern the work in this PR looks good then -- we should add a little more coverage to the test cases as mentioned and then we should be good to merge!

ravern commented:
Thanks @clarkzinzow @jaychia @samster25 for the feedback so far!

I've added more coverage to the test cases in terms of the values with different ways to double quote, and also renamed the option from `double_quote_escape` to `double_quote` according to what was suggested.

This PR should be ready for final review!
